File: Data Structures/Strings/caesarCipherEncryptor.js
// O(n) time | O(n) space
function caesarCipherEncryptor(string, key) {
  let alpabet = 'abcdefghijklmnopqrstuvwxyz';
  let alpabetLookUp = {
    a: 0,
    b: 1,
    c: 2,
    d: 3,
    e: 4,
    f: 5,
    g: 6,
    h: 7,
    i: 8,
    j: 9,
    k: 10,
    l: 11,
    m: 12,
    n: 13,
    o: 14,
    p: 15,
    q: 16,
    r: 17,
    s: 18,
    t: 19,
    u: 20,
    v: 21,
    w: 22,
    x: 23,
    y: 24,
    z: 25,
  };
  let currLetter;
  let stringIndex;
  let updatedIndex;
  let finalString = '';
  let newLetter;

  for (let i = 0; i < string.length; i++) {
    currLetter = string[i];
    stringIndex = alpabetLookUp[currLetter];
    updatedIndex = stringIndex + key;
    if (updatedIndex > 25) {
      updatedIndex = updatedIndex % 26;
    }
    newLetter = alpabet[updatedIndex];
    finalString = finalString.concat(newLetter);
  }
  return finalString;
}

File: Data Structures/Strings/caesarCipherEncryptor.py

######### PROBLEM ##########
# Given a non-empty string of lowercase letters and a non-negative integer
# representing a key, write a function that returns a new string obtained by
# shifting every letter in the input string by k positions in the alphabet,
# where k is the key.

# Note that letters should "wrap" around the alphabet; in other words, the
# letter z shifted by one returns the letter a


######### SOLUTION ##########

# Optimal Solution
# Time: O(n) where n is the length of the string
# Space: O(n)
def caesarCipherEncryptor(string, key):
    # Write your code here.
    alphabet = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l',
                'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
    finalString = ""
    for i in range(len(string)):
        letter = string[i]
        asciiValue = ord(letter)
        updatedAsciiValue = asciiValue + key
        newLetter = alphabet[(updatedAsciiValue-97) % 26]
        finalString += newLetter
    return finalString
